serve/server.go

```diff
@@ -75,14 +83,25 @@ func (s *Server) Serve(lis net.Listener) error {
 		grpc.UnknownServiceHandler(unknownHandler),
 	)
 	reflection.NewService(&s.files.Files).Register(s.gs)
-	return s.gs.Serve(lis)
+	return http.Serve(lis, h2c.NewHandler(s, &http2.Server{}))
+}
+
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	for _, method := range s.httpMethods {
+		if vars := httprule.MatchRequest(method.rule, r); vars != nil {
+			s.serveHTTPMethod(method, vars, w, r)
+			return
+		}
+	}
+	s.gs.ServeHTTP(w, r)
```

Comment:
Let's dispatch gRPC traffic first if the protocol version and content-type says so, so we don't need to iterate all the http methods before we even consider gRPC:
```
if r.ProtoMajor == 2 && strings.HasPrefix(r.Header.Get("Content-Type"), "application/grpc") {
    s.gs.ServeHTTP(w, r)
    return
}
```